Change default value for IPs

hemanthnakkina opened the pull request:
Currently the hook defaults the IP related
attributes to empty string. This resulted in
validation error when snap config get is mapped
to corresponding Pydantic model as model expects
IPv4/IPv6 address.

Change the default value to the IP address on the
default route.
Fix alias name for NetworkConfig ip_address.

hemanthnakkina committed:
Change default value for IPs

Currently the hook defaults the IP related
attributes to empty string. This resulted in
validation error when snap config get is mapped
to corresponding Pydantic model as model expects
IPv4/IPv6 address.

Change the default value to the IP address on the
default route.
Fix alias name for NetworkConfig ip_address.

javacruft commented:
I'd prefer to just default the node address - the others should use that if left unset, rather than having to set all three as soon as you wish to provide this value.

javacruft commented on openstack_hypervisor/hooks.py:
```diff
@@ -29,9 +29,11 @@
 from jinja2 import Environment, FileSystemLoader, Template
 from snaphelpers import Snap
 
+from openstack_hypervisor import utils
```
please can we stick to a single style for imports 

from openstack_hypervisor.utils import get_local_ip_by_default_route

also see note about naming of 'utils'

javacruft commented on openstack_hypervisor/hooks.py:
```diff
@@ -29,9 +29,11 @@
 from openstack_hypervisor.log import setup_logging
 
 UNSET = ""
+LOCAL_IP_BY_DEFAULT_ROUTE = utils.get_local_ip_by_default_route()
```
I'm not a huge fan of global execution in this way - see below

javacruft replied:
Actually see here - if we just pass the function as the value in the dict then in _update_default_config we can use something like

```
if callable(default):
  default = default()
```

this can be applied to socket.fqdn as well.

javacruft replied:
this avoids execution of functions during the construction of the dict as the module is imported

javacruft commented on openstack_hypervisor/utils.py:
```diff
@@ -0,0 +1,31 @@
+from netifaces import AF_INET, gateways, ifaddresses
+
+
+def get_local_ip_by_default_route() -> str:
```
please push this directly into the hooks module for the moment - we need to split that up but we can do that under a separate PR

hemanthnakkina committed:
Remove global variable LOCAL_IP_BY_DEFAULT_ROUTE

Remove global variable LOCAL_IP_BY_DEFAULT_ROUTE
and call the functions in DEFAULT_CONFIG dictionary
during install/configure methods.

javacruft commented on openstack_hypervisor/hooks.py:
```diff
@@ -164,6 +182,11 @@ def install(snap: Snap) -> None:
+            default = default()
+            DEFAULT_CONFIG[option] = default
     logging.info(f"Setting default config: {DEFAULT_CONFIG}")
     snap.config.set(DEFAULT_CONFIG)
```
I think we could possible switch this to use the _update_default_config function
